src/main/java/com/aws/greengrass/clientdevices/auth/util/WildcardTrie.java

```diff
@@ -0,0 +1,334 @@
+     *
+     * @param str string provided to get
+     */
+    static char getActualChar(String str) {
+        if (str.length() < 4) {
+            return nullChar;
+        }
+        // Match the escape format ${c}
+        if (str.charAt(0) == escapeChar && str.charAt(1) == '{' && str.charAt(3) == '}') {
+            return str.charAt(2);
+        }
+        return nullChar;
+    }
+
+    /**
+     * Match given string to the corresponding allowed resources trie. MQTT wildcards are not processed.
+     *
+     * @param str string to match.
+     */
+    @SuppressWarnings({"PMD.UselessParentheses", "PMD.CollapsibleIfStatements"})
+    public boolean matchesStandard(String str) {
+        if (str == null) {
+            return true;
+        }
+        if ((isWildcard && isTerminal) || (isTerminal && str.isEmpty())) {
+            return true;
+        }
+
+        boolean hasMatch = false;
+        Map<String, WildcardTrie> matchingChildren = new HashMap<>();
+        for (Map.Entry<String, WildcardTrie> e : children.entrySet()) {
+            // Succeed fast
+            if (hasMatch) {
+                return true;
+            }
+            String key = e.getKey();
+            WildcardTrie value = e.getValue();
+
+            // Process * wildcards
+            if (value.isWildcard && key.equals(GLOB_WILDCARD)) {
+                hasMatch = value.matchesStandard(str);
+                continue;
+            }
+
+            // Match normal characters
+            if (str.startsWith(key)) {
+                hasMatch = value.matchesStandard(str.substring(key.length()));
+                // Succeed fast
+                if (hasMatch) {
+                    return true;
+                }
+            }
+
+            // If I'm a wildcard, then I need to maybe chomp many characters to match my children
+            if (isWildcard) {
+                int foundChildIndex = str.indexOf(key);
+                int keyLength = key.length();
+                while (foundChildIndex >= 0) {
+                    matchingChildren.put(str.substring(foundChildIndex + keyLength), value);
+                    foundChildIndex = str.indexOf(key, foundChildIndex + 1);
+                }
+            }
+        }
+        // Succeed fast
+        if (hasMatch) {
+            return true;
+        }
+        if (isWildcard && !matchingChildren.isEmpty()) {
+            return matchingChildren.entrySet().stream().anyMatch((e) -> e.getValue().matchesStandard(e.getKey()));
+        }
+
+        return false;
+    }
+
+    /**
+     * Match given string to the corresponding allowed resources trie. MQTT wildcards are processed only if
+     * its a valid usage, otherwise treated as normal characters.
+     *
+     * @param str string to match
+     */
+    @SuppressWarnings({"PMD.UselessParentheses", "PMD.CollapsibleIfStatements"})
+    public boolean matchesMQTT(String str) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.